### ISSUE TITLE
RecursionError when reading an exported NWB file

### ISSUE THREAD

**Issue by kir0ul** [NeurodataWithoutBorders/pynwb#1301]
## Description
Hi,

I get the following error from `h5py` when I try to read a previously exported NWB file: `RecursionError: maximum recursion depth exceeded while calling a Python object`. It seems the NWB file gets corrupted during the export.

## Steps to Reproduce

This is an example NWB file I got from Suite2p: [ophys.zip](https://github.com/NeurodataWithoutBorders/pynwb/files/5285608/ophys.zip), which is needed unzipped on line 6 of the code snippet below.

```python
from pathlib import Path
import shutil
from pynwb import NWBHDF5IO

# File from Suite2p
nwb_path = Path("ophys.nwb")

# Create a backup file before it gets corrupted
backup_file = nwb_path.parent.joinpath("backup.nwb")
if not backup_file.exists():
    shutil.copy(nwb_path, backup_file)
shutil.copy(backup_file, nwb_path)

with NWBHDF5IO(str(nwb_path), "r") as read_io:
    nwbfile = read_io.read()
    img_seg = nwbfile.processing.get("ophys").data_interfaces.get("ImageSegmentation")
    PlaneSegmentationOld = img_seg.get_plane_segmentation("PlaneSegmentation")

    # Remove the old container and recreate a new one
    img_seg.plane_segmentations.pop("PlaneSegmentation")
    PlaneSegmentationNew = img_seg.create_plane_segmentation(
        description=PlaneSegmentationOld.description,
        imaging_plane=PlaneSegmentationOld.imaging_plane,
        name=PlaneSegmentationOld.name,
        reference_images=PlaneSegmentationOld.reference_images,
    )

    # Dummy data
    pix_mask = [[(0, 0, 1.1), (1, 1, 1.2), (2, 2, 1.3)], [(0, 0, 2.1), (1, 1, 2.2)]]
    iscell = [0, 1]

    # This is what I'm trying to replace in the old file but it crashes without this code block
    for elem in pix_mask:
        PlaneSegmentationNew.add_roi(pixel_mask=elem)
    # PlaneSegmentationOld["iscell"][()] = iscell  # <-- Doesn't work
    PlaneSegmentationNew.add_column("iscell", "two columns - iscell & probcell", iscell)

    # Export the file to a new location
    export_filename = nwb_path.parent.joinpath("tmp.nwb")
    with NWBHDF5IO(str(export_filename), mode="w") as export_io:
        export_io.export(src_io=read_io, nwbfile=nwbfile)

# Remove the old file and replace it with the new one
nwb_path.unlink()
export_filename.rename(nwb_path)

# Try to read the file but it seems corrupted
with NWBHDF5IO(str(nwb_path), "r") as read_io:
    nwbfile = read_io.read()  # <-- This is where it should crash
    print(nwbfile)
```

## Environment

    Python Executable: Conda
    Python Version: Python 3.7.9
    Operating System: Linux
    HDMF Version: 2.2.0
    Version of PyNWB used: 1.4.0

## Checklist

- [X] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues)?
- [X] Have you included a brief and descriptive title?
- [X] Have you included a clear description of the problem you are trying to solve?
- [X] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?

**Comment by bendichter** [comment]
Thanks for the report, @kir0ul! I was able to reproduce your error. Thank you for supplying a full self-contained script. I was also able to reduce your code down to the following and get the same error:

```python
from pathlib import Path
import shutil
from pynwb import NWBHDF5IO

# File from Suite2p
nwb_path = Path("/Volumes/easystore5T/data/misc/ophys.nwb")

with NWBHDF5IO(str(nwb_path), "r") as read_io:
    nwbfile = read_io.read()
    img_seg = nwbfile.processing.get("ophys").data_interfaces.get("ImageSegmentation")
    img_seg.plane_segmentations.pop("PlaneSegmentation")

    # Export the file to a new location
    export_filename = nwb_path.parent.joinpath("tmp.nwb")
    with NWBHDF5IO(str(export_filename), mode="w") as export_io:
        export_io.export(src_io=read_io, nwbfile=nwbfile)

# Try to read the file but it seems corrupted
with NWBHDF5IO(str(export_filename), "r") as read_io:
    nwbfile = read_io.read()  # <-- This is where it should crash
    print(nwbfile)
```

@rly, you have been deep in this type of stuff. What is the proper way to alter a pre-existing file?

Also, we should try to put up a safe-guard to prevent this type of issue in the future. I'd much rather fail on write than write a file that we can't read.

**Comment by kir0ul** [comment]
Update: this doesn't crash anymore with HDMF 2.4.0!

**Comment by DaveOC90** [comment]
Hi,

Sorry to dredge this up, but I recently ran into this error, and am unsure how to proceed. For the record, I replicated what @kir0ul [kir0ul](https://github.com/kir0ul) did at the top of the thread, without issue.

However the error was raised for when I tried to iterate over a list of images, load the data, and add the data to an nwb file. I am using newer versions of pynwb and hdmf than specified above. I have come up with a simple test case that one should be able to copy and paste which raises the error on my machine. I instantiate an nwbfile, and save it. I then enter a for loop which was intended to recursively load image data, and add another TwoPhotonSeries to the nwbfile, without keeping more than one image in memory. The error occurs in the second loop, when trying to load the nwbfile. See below: 

```
from datetime import datetime
from dateutil.tz import tzlocal

import numpy as np
from pynwb import NWBFile, TimeSeries, NWBHDF5IO
from pynwb.file import Subject
from pynwb.image import ImageSeries
from pynwb.ophys import TwoPhotonSeries, OpticalChannel, ImageSegmentation, \
    Fluorescence, CorrectedImageStack, MotionCorrection, RoiResponseSeries

subId,seshId = 'sub-01','ses-1'

sub = Subject(subject_id = subId,species='Mus musculus')

nwbObj = NWBFile(
    'Dual Wavelength Mesoscale Calcium Imaging of Multiple Neuron Types',
    subId+'_'+seshId,
    datetime.now(tzlocal()),
    experimenter='',
    lab='TBD',
    institution='TBD',
    experiment_description='Mesoscale calcium one photon imaging of awake and anesthetized SLC neurons in mouse isocortex'
                           'and awake imaging of PV, SOM, and VIP neurons in mouse isocortex',
    session_id=seshId,
    subject = sub)


device = nwbObj.create_device(
    name="Microscope",
    description="Zeiss AxioZoom v.16 microscope with a PlanNeoFluar Z 1x/0.25 objective",
    manufacturer="Zeiss")

nwbFilePath = subId+'_'+seshId+'.nwb'
with NWBHDF5IO(nwbFilePath, 'w') as writeio:
    writeio.write(nwbObj)

for i in range(0,2):

    with NWBHDF5IO(nwbFilePath, mode = 'r+') as io:
        nwbObjIter = io.read()

        imgData = np.zeros([3,3,3])

        optical_channel_uv = OpticalChannel(
            name="OpticalChannel",
            description="Uv Wavelength",
            emission_lambda=395.)

        imaging_plane_uv = nwbObjIter.create_imaging_plane(
            name="ImagingPlane",
            optical_channel=optical_channel_uv,
            imaging_rate=10.,
            description="Whole cortex",
            device=device,
            excitation_lambda=395.,
            indicator="GCamp6",
            location="Whole cortex",
            grid_spacing=[.25, .25],
            grid_spacing_unit="millimeters")

        image_series_uv = TwoPhotonSeries(
            name='onePhotonSeries',
            data=imgData,
            imaging_plane=imaging_plane_uv,
            rate=10.0,
            description='One photon uv wavelength data',
            unit = 'N/A')

        nwbObjIter.add_acquisition(image_series_uv)

        io.write(nwbObjIter)

```


## Traceback:

```
RecursionError                            Traceback (most recent call last)
<ipython-input-1-e3a03c29cbcf> in <module>
     39 
     40     with NWBHDF5IO(nwbFilePath, mode = 'r+') as io:
---> 41         nwbObjIter = io.read()
     42 
     43         imgData = np.zeros([3,3,3])

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py in read(self, **kwargs)
    496                                        % (self.source, self.__mode))
    497         try:
--> 498             return call_docval_func(super().read, kwargs)
    499         except UnsupportedOperation as e:
    500             if str(e) == 'Cannot build data. There are no values.':  # pragma: no cover

~/anaconda3/lib/python3.7/site-packages/hdmf/utils.py in call_docval_func(func, kwargs)
    422 def call_docval_func(func, kwargs):
    423     fargs, fkwargs = fmt_docval_args(func, kwargs)
--> 424     return func(*fargs, **fkwargs)
    425 
    426 

~/anaconda3/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    581             def func_call(*args, **kwargs):
    582                 pargs = _check_args(args, kwargs)
--> 583                 return func(args[0], **pargs)
    584         else:
    585             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/io.py in read(self, **kwargs)
     35     def read(self, **kwargs):
     36         """Read a container from the IO source."""
---> 37         f_builder = self.read_builder()
     38         if all(len(v) == 0 for v in f_builder.values()):
     39             # TODO also check that the keys are appropriate. print a better error message

~/anaconda3/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    581             def func_call(*args, **kwargs):
    582                 pargs = _check_args(args, kwargs)
--> 583                 return func(args[0], **pargs)
    584         else:
    585             def func_call(*args, **kwargs):

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py in read_builder(self)
    513             ignore.add(self.__file[specloc].name)
    514         if f_builder is None:
--> 515             f_builder = self.__read_group(self.__file, ROOT_NAME, ignore=ignore)
    516             self.__read[self.__file] = f_builder
    517         return f_builder

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py in __read_group(self, h5obj, name, ignore)
    654                         obj_type = kwargs['groups']
    655                     if builder is None:
--> 656                         builder = read_method(sub_h5obj)
    657                         self.__set_built(sub_h5obj.file.filename, sub_h5obj.id, builder)
    658                     obj_type[builder.name] = builder

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py in __read_group(self, h5obj, name, ignore)
    654                         obj_type = kwargs['groups']
    655                     if builder is None:
--> 656                         builder = read_method(sub_h5obj)
    657                         self.__set_built(sub_h5obj.file.filename, sub_h5obj.id, builder)
    658                     obj_type[builder.name] = builder

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py in __read_group(self, h5obj, name, ignore)
    635                             builder = self.__read_dataset(target_obj, builder_name)
    636                         else:
--> 637                             builder = self.__read_group(target_obj, builder_name, ignore=ignore)
    638                         self.__set_built(sub_h5obj.file.filename,  target_obj.id, builder)
    639                     builder.location = parent_loc

... last 3 frames repeated, from the frame below ...

~/anaconda3/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py in __read_group(self, h5obj, name, ignore)
    635                             builder = self.__read_dataset(target_obj, builder_name)
    636                         else:
--> 637                             builder = self.__read_group(target_obj, builder_name, ignore=ignore)
    638                         self.__set_built(sub_h5obj.file.filename,  target_obj.id, builder)
    639                     builder.location = parent_loc

RecursionError: maximum recursion depth exceeded while calling a Python object

```

## Environment
```
Python Executable: Conda
Python Version: Python 3.7.11
Operating System: CentOS
HDMF Version: 3.2.1
Version of PyNWB used: 2.0.1
```

**Comment by bendichter** [comment]
```python
image_series_uv = TwoPhotonSeries(
            name='onePhotonSeries',
            data=imgData,
            imaging_plane=imaging_plane_uv,
            rate=10.0,
            description='One photon uv wavelength data',
            unit = 'N/A')

        nwbObjIter.add_acquisition(image_series_uv)
```

I am not sure if this is what is causing your error (if it is we need to fix the error message) but I would expect this code to fail. HDF5 is like a file system. In a file system you can't have two folders with the same name in the same folder. Similarly with HDF5, you can't have two or more Groups called "onePhotonSeries" in the `/acquisitions` Group. You have the same problem for the name of the ImagingPlane. The OpticalChannel objects actually do not need unique names since they are stored within the ImagingPlane objects.

**Comment by DaveOC90** [comment]
Hi Ben,

Thanks for your response. I added in an iterative variable to update the two photon series name, but still got the error. I then updated the imaging_plane line to reference the imaging_plane from the newly loaded nwb object. That is:

```
        image_series = TwoPhotonSeries(
            name='onePhotonSeries'+str(i+1),
            data=imgData,
            imaging_plane= nwbObjIter.get_imaging_plane(),
            rate=10.0,
            description='One photon uv wavelength data',
            unit = 'Amplitude')

```

Either the latter or both of these changes fixed the problem. I am still getting to grips with nwb/hdf5 but what you say makes sense, and thanks again for responding. I appreciate that the above code should have errored.
